### PR TITLE
[TRTLLM-11733][perf] Cache constant text computations across denoise steps in LTX2

### PR DESCRIPTION
**Pull request by luyiyun1021**
## Summary

Cache text-derived computations that are constant across denoise steps in the LTX2 pipeline.

LTX2 uses text context (prompt embeddings) that is **constant throughout the entire denoising loop** — the same prompt produces the same text embedding every step. Currently, several expensive operations on this constant input are redundantly recomputed at every step:

1. **Preprocessor level**: `caption_projection(context)` (Linear GEMM), attention mask construction, and RoPE positional embeddings
2. **Block level**: K/V projections (`to_k`, `to_v`) and K-normalization (`norm_k`) in 48 layers × 2 modalities (video + audio) of text cross-attention (`attn2` / `audio_attn2`)

Together these produce ~678 redundant kernels per step.

### Approach

**Two-level caching:**

1. **Preprocessor caching** (`transformer_args.py`): Cache `caption_projection(context)`, `attention_mask`, and RoPE across denoise steps using `data_ptr()` as cache key. Runs outside `torch.compile` (in `LTXModel.forward()`), so `data_ptr()` is safe.

2. **Block-level KV cache** (`transformer_ltx2.py`): Each `BasicAVTransformerBlock` stores cached KV as plain tuple attributes (`_text_kv_video`, `_text_kv_audio`). On first use per generate() call (`is None` check), `project_kv()` fills the cache. Subsequent denoise steps reuse it via `pre_projected_kv=` parameter.

3. **Cache lifecycle** (`pipeline_ltx2.py`): `invalidate_text_kv_cache()` resets both block-level KV caches and preprocessor caches before each denoising loop, preventing stale `data_ptr()` hits across `generate()` calls.

4. **Two-stage pipeline support** (`pipeline_ltx2_two_stages.py`): Stage 2 uses LoRA-merged weights for refinement denoising. `invalidate()` is called after LoRA merge so the cache refills with the merged weights on the first Stage 2 step. Stage 2 only uses `CfgPass.COND` (no CFG). NVTX markers (`refinement_step {i}`) are added to the stage 2 denoise loop for profiling.

5. **Encoder output caching** (`pipeline_ltx2.py`, `pipeline_ltx2_two_stages.py`): Text encoder output (Gemma3 + Connector) is cached in `LTX2TextContextCache` after Stage 1 and reused in Stage 2, skipping redundant `_encode_prompt()` + `_process_connectors()` calls. These outputs are prompt-only dependent — not affected by resolution or LoRA weights — so they survive `invalidate()`. Saves ~27ms GPU time per two-stage run.

**torch.compile compatibility**: All cache logic (fill, invalidate, store/get) runs outside compiled blocks in `LTXModel.forward()`. Compiled blocks only receive pre-projected KV as parameters — no conditional logic, no guards, no recompilation.

### Single-Stage 40-Step Performance (NVFP4, B200, 736×1280, 121 frames, torch.compile ON)

All measurements on same node (umbriel-b200-048). Steady-state runs (excludes first-run recompile overhead from unrelated NVFP4 dtype guards).

**Step denoising time (avg of steps 2–40):**

| GPU | Baseline | Optimized | Δ | Δ % |
|-----|----------|-----------|---|-----|
| 1GPU | 710ms | 702ms | -8ms | **-1.2%** |
| 2GPU | 452ms | 446ms | -6ms | **-1.3%** |
| 4GPU | 270ms | 266ms | -4ms | **-1.5%** |
| 8GPU | 190ms | 180ms | -10ms | **-5.3%** |

**E2E (total generate() wall time):**

| GPU | Baseline | Optimized | Δ | Δ % |
|-----|----------|-----------|---|-----|
| 1GPU | 30.89s | 30.57s | -0.32s | **-1.0%** |
| 2GPU | 20.60s | 20.32s | -0.29s | **-1.4%** |
| 4GPU | 13.34s | 13.11s | -0.23s | **-1.7%** |
| 8GPU | 9.94s | 9.69s | -0.25s | **-2.5%** |

### Two-Stage 40-Step Performance (NVFP4, B200, 768×1280, 121 frames, 40+3 steps)

Two-stage pipeline: stage 1 at half resolution (384×640, 40 steps) + LoRA merge + spatial upsample + stage 2 at full resolution (768×1280, 3 steps). Same code, cache disabled via env var for baseline.

**E2E (total generate() wall time, 1GPU):**

| | Baseline | Optimized | Δ | Δ % |
|--|----------|-----------|---|-----|
| 1GPU | 15.233s | 14.884s | -0.35s | **-2.3%** |

### 1GPU Kernel-Level Breakdown (nsys, per step, verified match=Y)

Kernel sequences verified identical via `diff` on stripped triton kernel names. Two-pointer alignment confirms exact match.

**Single-stage (denoise_step 2, 736×1280, NVFP4): 678 kernels eliminated per step**

| What was eliminated | Kernels saved | GPU time saved |
|---------------------|:---:|---:|
| GEMM (nvjet, KV projection) | 196 | 2.13ms |
| QK RMSNorm (K normalization) | 96 | 1.03ms |
| FP4 Quantize (KV quantization) | 168 | 0.99ms |
| RMSNorm (FlashInfer, context norm) | 168 | 0.65ms |
| BF16 GEMM (excluded layers) | 28 | 0.56ms |
| Elementwise | 47 | 0.55ms |
| CatArrayBatchedCopy | 3 | 0.12ms |
| **Total** | **678** | **~6.0ms** |

**Two-stage Stage 1 (denoise_step 2, 384×640): 678 kernels eliminated per step**

| What was eliminated | Kernels saved | GPU time saved |
|---------------------|:---:|---:|
| GEMM (nvjet, KV projection) | 196 | 1.76ms |
| QK RMSNorm (K normalization) | 96 | 0.85ms |
| FP4 Quantize (KV quantization) | 168 | 0.79ms |
| RMSNorm (FlashInfer, context norm) | 168 | 0.53ms |
| Elementwise | 47 | 0.17ms |
| CatArrayBatchedCopy | 3 | 0.03ms |
| **Total** | **678** | **4.13ms** |

**Stage 2 (refinement_step 1, 768×1280): 678 kernels eliminated per step**

| What was eliminated | Kernels saved | GPU time saved |
|---------------------|:---:|---:|
| GEMM (nvjet, KV projection) | 196 | 2.13ms |
| QK RMSNorm (K normalization) | 96 | 1.03ms |
| FP4 Quantize (KV quantization) | 168 | 0.99ms |
| RMSNorm (FlashInfer, context norm) | 168 | 0.65ms |
| Elementwise | 47 | 0.58ms |
| CatArrayBatchedCopy | 3 | 0.13ms |
| **Total** | **678** | **5.50ms** |

Note: Stage 2 kernels are slower per-kernel than Stage 1 due to higher resolution (768×1280 vs 384×640).

### Tests

- Unit test covering cache numerical equivalence (bitwise identical), invalidation correctness (different context → different output), CFG slot isolation, and single-GPU CFG simulation
- All existing LTX2 tests pass (12/12)

## PR Checklist

- [x] Please check this if you have read the [contribution guidelines](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by zhenhuaw-me — COMMENTED** [review]
Thanks! I think in general it's a good finding. We can follow up to polish the implementation.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                k, v = block.attn2.project_kv(video_context)
+                self._store(self._video_kv, s, layer_idx, k, v)
+            if audio_context is not None:
+                k, v = block.audio_attn2.project_kv(audio_context)
+                self._store(self._audio_kv, s, layer_idx, k, v)
+
+        self._kv_dirty[s] = False
+
+    def get_kv(
+        self, modality: str, is_unconditional: bool, layer_idx: int
```
[the remark this answers is not in the document]
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py]: Agreed. Changed `modality: str` to `ModalityType(IntEnum)` with `.VIDEO` and `.AUDIO` members. Used as list index internally. See latest commit.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        audio_context: torch.Tensor | None,
+    ) -> None:
+        """Fill per-block KV cache if dirty.  Runs outside torch.compile."""
+        s = _slot(is_unconditional)
+        if not self._kv_dirty[s]:
+            return
+
+        for layer_idx, block in enumerate(blocks):
+            if video_context is not None:
+                k, v = block.attn2.project_kv(video_context)
```
[the remark this answers is not in the document]
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py]: Good catch. Moved `project_kv()` calls out of the cache into `LTXModel.forward()`. Cache now only does pure store/get via `store_kv()` and `get_kv()` — no computation or block references. See latest commit.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/pipeline_ltx2.py
[the remark this answers is not in the document]
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/pipeline_ltx2.py]: Good point. Stage 1 automatically benefits via inheritance. Stage 2 currently calls `self.transformer()` without `text_cache` (line 940 in `_refinement_denoise`), so it does not use the cache yet. Adding `text_cache` + `invalidate()` before Stage 2 denoise loop so the 3 refinement steps also benefit. See next commit.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py
[the remark this answers is not in the document]
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/text_context_cache.py]: Correct — after LoRA merge, KV projections change. The fix is straightforward: call `self._text_cache.invalidate()` after `_apply_lora_deltas()` and pass `text_cache=self._text_cache, cfg_pass=CfgPass.COND` into the Stage 2 transformer calls. This forces a refill with the merged weights. No new cache instance needed — `invalidate()` marks slots dirty and the first step refills the existing buffers via `copy_()`. See next commit.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/ltx2_core/transformer_args.py [hunk 3]
```diff
@@ -125,21 +135,32 @@ def _prepare_positional_embeddings(
...
-            inner_dim=self.inner_dim,
-            max_pos=self.max_pos,
-            use_middle_indices_grid=self.use_middle_indices_grid,
-            num_attention_heads=self.num_attention_heads,
-            x_dtype=modality.latent.dtype,
-        )
+
+        # Context projection, attention mask, and RoPE are constant across
+        # denoise steps — cached per CFG slot and reused.
+        entry = self._text_cache.get_preproc(cfg_pass, self._modality)
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/models/ltx2/ltx2_core/transformer_args.py]: if entry is always not None?

**Review by zhenhuaw-me — APPROVED** [review]
LGTM. Thanks!

File: tensorrt_llm/_torch/visual_gen/models/ltx2/ltx2_core/transformer_args.py (hunk 3 above)
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/ltx2_core/transformer_args.py]: Yes

**Review by yibinl-nvidia — APPROVED** [review]
LGTM to the two stage pipeline changes.
